### DISCUSSION TITLE
Exercises Vision Document

### DISCUSSION THREAD

**Discussion by iK4tsu** [iK4tsu/dlings#4] [Ideas]
# VISION DOCUMENT
This is a list of on-going thoughts of how the exercises should be organized. Not everything in the language, from language features to the standard library, will be mentioned in the main set of exercises. The language is super extensive and the more language features and standard library support we put on the main set, the longer and complex it will be. Some who's starting programming and/or learning D does not need to know everything about the language at once to start creating something in D.

My idea is to have a main set that focuses only in the absolute essential, and provide a mechanism in dlings to explore other functionalities either native or provided by the standard library. Having a way to split exercises into multiple groups also makes it possible to extend dlings to other knowledge levels without forcing the resolution of known things. I believe there is always something in D that is new to someone, no matter the level knowledge and time that person has with the language. By creating these groups, we open a door to multiple topics, having tips and tricks, features not really mentioned in the spec, common pitfalls and how to overcome them, explain what some previews do, go in depth with dip1000 and in, etc.

Below there are some thoughts that will be part of the main set, some loose thought of things to talk either in the main set or other category and some categories I think make sense to have.

## Structured Thoughts
- [ ] Introduction
  - [ ] Welcome to dlings
  - [ ] Hello World
- [ ] Variables
  - [ ] Declaration
  - [ ] Default initializer
  - [ ] Immutability
- [ ] Floating Point
  - [ ] Nan
  - [ ] Infinity
- [ ] Functions
  - [ ] Normal
  - [ ] Templated
- [ ] If
  - [ ] If
  - [ ] Else
  - [ ] Else If
  - [ ] Ternary Operator
- [ ] Quiz
- [ ] Arrays ( Static, Dynamic, Slices and Associative Arrays )
  - [ ] Static Arrays
  - [ ] Dynamic Arrays
  - [ ] Slices
  - [ ] Assoaciative Arrays
- [ ] Loops
  - [ ] While
  - [ ] For
  - [ ] Foreach
  - [ ] Do While
- [ ] Quiz 
- [ ] Enum
- [ ] Switch Case
  - [ ] Default
  - [ ] Final Switch
- [ ] Type Properties
  - [ ] Integer
  - [ ] Floating Point
  - [ ] Slices
  - [ ] Associative Arrays

## Loose thoughts
- [ ] Import
- [ ] Shared
- [ ] Exceptions
- [ ] Templates
- [ ] Meta
- [ ] Contracts In | Out | Invariant
- [ ] Unit Testing
- [ ] Lazy
- [ ] Variadic Arguments
- [ ] Ref Out In
- [ ] Auto Ref
- [ ] Lambdas
- [ ] Scope Exit | Success | Failure
- [ ] Classes
- [ ] Interfaces
- [ ] Inheritance
- [ ] UFCS
- [ ] Pragma
- [ ] Extern
- [ ] Alias This
- [ ] Alias
- [ ] With
- [ ] Pointers
- [ ] Union
- [ ] Struct
- [ ] Goto & Labels
- [ ] Is Expressions
- [ ] Ranges
- [ ] Mixins
- [ ] UDA
- [ ] Parallelism
- [ ] Mixin Template

### Did you know
- [ ] Reverse Foreach
- [ ] Tuple Unpacking
- [ ] Anonymous Classes
- [ ] Shortned Methods

### Multithread
- [ ] Atomic Operations
- [ ] Concurrency
- [ ] Message Passing
- [ ] Fibers

### TBD

**Comment by WebFreak001** [comment]
you could integrate exercises from https://tour.dlang.org into this

**Reply by iK4tsu** [reply]
I'm basing some exercises on D Tour and have links to some D Tour topics in each chapter README when applied. Or am I misunderstanding what you meant?

**Reply by WebFreak001** [reply]
no, that's actually exactly what I meant, but didn't phrase exactly like that ^^
